pkg/util/push/push.go

```diff
@@ -85,43 +84,38 @@ func handler(maxRecvMsgSize int,
 				logger = log.WithSourceIPs(source, logger)
 			}
 		}
-		bufHolder := bufferPool.Get().(*bufHolder)
-		var req mimirpb.PreallocWriteRequest
-		buf, err := parser(ctx, r, maxRecvMsgSize, bufHolder.buf, &req)
-		if err != nil {
-			level.Error(logger).Log("err", err.Error())
-
-			// Check for httpgrpc error.
-			if resp, ok := httpgrpc.HTTPResponseFromError(err); ok {
-				http.Error(w, string(resp.Body), int(resp.Code))
-			} else {
-				http.Error(w, err.Error(), http.StatusBadRequest)
+		supplier := func() (*mimirpb.WriteRequest, func(), error) {
+			bufHolder := bufferPool.Get().(*bufHolder)
+			var req mimirpb.PreallocWriteRequest
+			buf, err := parser(ctx, r, maxRecvMsgSize, bufHolder.buf, &req)
+			if err != nil {
+				// Check for httpgrpc error, default to client error if parsing failed
+				if _, ok := httpgrpc.HTTPResponseFromError(err); !ok {
+					err = httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+				}
+
+				bufferPool.Put(bufHolder)
+				return nil, nil, err
+			}
+			// If decoding allocated a bigger buffer, put that one back in the pool.
+			if buf = buf[:cap(buf)]; len(buf) > len(bufHolder.buf) {
+				bufHolder.buf = buf
 			}
```

Comment:
would this not be the same and more similar to similar to other typical usages of slices, like for example `buf = append(buf, ...)`?

```suggestion
			var err error
			bufHolder.buf, err = parser(ctx, r, maxRecvMsgSize, bufHolder.buf, &req)
			bufHolder.buf = bufHolder.buf[:cap(bufHolder.buf)]
			if err != nil {
				// Check for httpgrpc error, default to client error if parsing failed
				if _, ok := httpgrpc.HTTPResponseFromError(err); !ok {
					err = httpgrpc.Errorf(http.StatusBadRequest, err.Error())
				}

				bufferPool.Put(bufHolder)
				return nil, nil, err
			}
```

Reply:
so I wrote this 6c8ea108bca570bd071be94d3e325a0fde45f5b1, but it has the caveat that if the returned `buf` is indeed a different slice, then we retain both slices until the cleanup is called - at the end of the request - instead of during the request parsing.

Reply:
because of this I reverted the commit in https://github.com/grafana/mimir/pull/2419/commits/ceaab27804030e92eefc940c96f8f6dfaa1ee584